File: java/src/test/java/app/endpoints/APIendpoints.java
package app.endpoints;

import static io.gatling.javaapi.core.CoreDsl.*;
import static io.gatling.javaapi.http.HttpDsl.*;

import io.gatling.javaapi.http.*;

public class APIendpoints {

  public static final HttpProtocolBuilder withAuthenticationHeader(
      HttpProtocolBuilder protocolBuilder) {
    return protocolBuilder.header(
        "Authorization",
        session -> session.contains("accessToken") ? session.getString("accessToken") : "");
  }

  public static final HttpRequestActionBuilder session =
      http("Session")
          .get("/session")
          .check(status().is(200))
          .check(jmesPath("sessionId").saveAs("sessionId"));

  public static final HttpRequestActionBuilder products =
      http("Product page: #{pageIndex}")
          .get("/products")
          .queryParam("page", "#{pageIndex}")
          .check(status().is(200))
          .check(jmesPath("products").saveAs("products"));

  public static final HttpRequestActionBuilder login =
      http("Login")
          .post("/login")
          .asFormUrlEncoded()
          .formParam("username", "#{username}")
          .formParam("password", "#{password}")
          .check(status().is(200))
          .check(jmesPath("accessToken").saveAs("accessToken"));

  public static final HttpRequestActionBuilder search =
      http("Search").get("/products").queryParam("search", "t-shirt").check(status().is(200));

  public static final HttpRequestActionBuilder cart =
      http("Add to Cart")
          .post("/cart")
          .asJson()
          .body(ElFileBody("bodies/cart.json"))
          .check(status().is(200));

  public static final HttpRequestActionBuilder checkOut =
      http("Checkout")
          .post("/checkout")
          .asJson()
          .body(ElFileBody("bodies/cart.json"))
          .check(status().is(200));
}

File: scala/src/test/scala/app/endpoints/APIendpoints.scala
package app.endpoints

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import io.gatling.http.protocol.HttpProtocolBuilder

object APIendpoints {

  def withAuthenticationHeader(protocolBuilder: HttpProtocolBuilder): HttpProtocolBuilder = {
    protocolBuilder.header(
      "Authorization",
      session => if (session.contains("accessToken")) session("accessToken").as[String] else ""
    )
  }

  val session = http("Session")
    .get("/session")
    .check(status.is(200))
    .check(jmesPath("sessionId").saveAs("sessionId"))

  val products = http("Product page: #{pageIndex}")
    .get("/products")
    .queryParam("page", "#{pageIndex}")
    .check(status.is(200))
    .check(jmesPath("products").saveAs("products"))

  val login = http("Login")
    .post("/login")
    .formParam("username", "#{username}")
    .formParam("password", "#{password}")
    .check(status.is(200))
    .check(jmesPath("accessToken").saveAs("accessToken"))

  val search = http("Search")
    .get("/products")
    .queryParam("search", "t-shirt")
    .check(status.is(200))

  val cart = http("Add to Cart")
    .post("/cart")
    .asJson
    .body(ElFileBody("bodies/cart.json"))
    .check(status.is(200))

  val checkOut = http("Checkout")
    .post("/checkout")
    .asJson
    .body(ElFileBody("bodies/cart.json"))
    .check(status.is(200))
}
